3255 sync ethereum transaction confirmation with backend

DmitriyNoa opened the pull request.

arjunu commented on packages/web/app/components/shared/utils.ts:
```diff
@@ -45,6 +49,77 @@ const formatEuroValueToString = (value: string) =>
+const getPendingTransactionType = (
+  transaction: TxPendingWithMetadata,
+):
+  | ETransactionType.ETO_INVESTMENT
```
what about `typeof IPendingTransaction.type` ?

Moejoe90 commented on packages/web/app/modules/tx/event-channel/sagas.ts:
```diff
@@ -38,8 +39,19 @@ export function* getTransactionOrThrow(
 
+  const blockHeight = process.env.NF_SUCCESSFUL_TRANSACTION_BLOCK_HEIGHT
+    ? parseInt(process.env.NF_SUCCESSFUL_TRANSACTION_BLOCK_HEIGHT, 10)
+    : 3;
```
default should be `0` 
Make zero a constant with `NO_BLOCK_HIGHT` as a name

DmitriyNoa replied:
Done

Moejoe90 commented on packages/web/app/modules/tx/event-channel/sagas.ts:
```diff
@@ -38,8 +39,19 @@ export function* getTransactionOrThrow(
+    txReceipt &&
+    txReceipt.blockNumber &&
+    currentBlockNumber - tx.blockNumber >= blockHeight &&
+    currentBlockNumber - txReceipt.blockNumber >= blockHeight;
```
I think this logic now is becoming more complex, 
lets abstarct this to a util function that checks the block hight this includes taking the data from `process.env`
We can even put it as part of the `shared/package`

DmitriyNoa replied:
Moved

Moejoe90 commented on packages/web/app/lib/api/users/interfaces.ts:
```diff
@@ -40,6 +44,24 @@ export interface IUserInput {
+    | ETransactionType.ETO_INVESTMENT
+    | ETransactionType.NEUR_REDEEM
+    | ETransactionType.ETO_TOKENS_CLAIM
+    | ETransactionType.TRANSFER;
```
why are these not all into one `type` ? where 
```
type something = ETransactionType.ETO_INVESTMENT
    | ETransactionType.NEUR_REDEEM
    | ETransactionType.ETO_TOKENS_CLAIM
    | ETransactionType.TRANSFER;
```

DmitriyNoa replied:
Moved

Moejoe90 commented on packages/web/app/lib/api/users/interfaces.ts:
```diff
@@ -40,6 +44,24 @@ export interface IUserInput {
   walletSubtype: EWalletSubType;
 }
 
+export interface IPendingTransaction {
```
This is noy the right way to create this interface each transaction has its own `PendingTransaction` TYPE 

You don't need to create this new generic type as its full of `optional` values now for a good reason have a look at `TWithdrawAdditionalData` or  `TInvestmentAdditionalData` etc and how they are combined

DmitriyNoa replied:
This is the problem. There many cases in which pending transactions can be, depending on a type of transaction. Sometimes we have currency symbols, some transaction doesn't have those. The same goes with decimals some don't.  The data is very inconsistent so it's difficult to come up with a type. We can check this together

DmitriyNoa replied:
Moved to already existing TxPendingWithMetadata type.

Moejoe90 commented on packages/web/app/modules/tx/monitor/selectors.ts:
```diff
@@ -14,6 +27,37 @@ export const selectPlatformPendingTransaction = (
 ): TxPendingWithMetadata | undefined =>
   state.txMonitor.txs.pendingTransaction as TxPendingWithMetadata | undefined;
 
+export const selectPlatformMiningTransaction = (state: IAppState): IPendingTransaction | null => {
```
I think putting `getPendingTransactionAmount` `getPendingTransactionCurrency` and `getPendingTransactionType` in a `utils.ts` file here makes sense as they are the main use case for this. 

Also btw when creating the correct type for each transaction you won't need to have this complicated selector as each type should return the correct data

DmitriyNoa replied:
Moved to tx utils

DmitriyNoa commented on packages/web/app/components/modals/tx-sender/investment-flow/InvestmentTransactionDetails.spec.tsx:
```diff
@@ -24,6 +24,7 @@ describe("InvestmentTransactionDetails", () => {
         equityTokenName: "Quintessence",
       },
     },
+    investmentType: "REDEEM_TOKEN",
```
Done

DmitriyNoa commented on packages/web/app/components/modals/tx-sender/investment-flow/Success.stories.tsx:
```diff
@@ -31,6 +31,7 @@ const props = {
         equityTokenName: "Quintessence",
       },
     },
+    investmentType: "REDEEM_TOKEN",
```
Done

DmitriyNoa commented on packages/web/app/components/modals/tx-sender/shared/TxError.stories.tsx:
```diff
@@ -61,6 +61,7 @@ const investTxData: React.ComponentProps<typeof TxErrorLayout> = {
         equityTokenName: "Quintessence",
       },
     },
+    investmentType: "REDEEM_TOKEN",
```
Done

DmitriyNoa commented on packages/web/app/components/modals/tx-sender/investment-flow/Summary.stories.tsx:
```diff
@@ -20,6 +20,7 @@ const data = {
         equityTokenName: "Quintessence",
       },
     },
+    investmentType: "REDEEM_TOKEN",
```
Done

DmitriyNoa commented on packages/web/app/components/modals/tx-sender/shared/TxSuccess.stories.tsx:
```diff
@@ -22,6 +22,7 @@ const txData: React.ComponentProps<typeof TxSuccessLayout> = {
         equityTokenName: "Quintessence",
       },
     },
+    investmentType: "REDEEM_TOKEN",
```
Done

Moejoe90 commented on packages/web/app/components/shared/transaction/Transaction.stories.tsx:
```diff
@@ -0,0 +1,82 @@
+import { Transaction } from "./Transaction";
+
+storiesOf("Transaction", module)
+  .add("default", () => {
```
this is still not addresses

DmitriyNoa commented on packages/web/app/components/shared/transaction/Transaction.tsx:
```diff
@@ -0,0 +1,50 @@
+      <div className={styles.info}>
+        {icon && <div className={styles.icon}>{icon}</div>}
+        <TransactionData
+          top={<TransactionName transaction={transaction} />}
```
It requires more data, transactionDirection, type, currency, amount as it computes the name from those

DmitriyNoa commented on packages/web/app/components/shared/utils.ts:
```diff
@@ -45,6 +49,77 @@ const formatEuroValueToString = (value: string) =>
+        amount = transaction.transactionAdditionalData.investmentEth;
+    }
+
+    if (transaction.transactionAdditionalData.tokenQuantity) {
```
For some transactions amount is missing so I have to take it from tokenQuantity

DmitriyNoa commented on packages/web/app/test-e2e/pending-transactions/generalPendingTxFixture.ts:
```diff
@@ -33,6 +33,7 @@ export const generalPendingTxFixture = (
     tokenSymbol: "QTT" as EquityToken,
     tokenImage: "test",
     tokenDecimals: 18,
+    investmentType: "NEUR",
```
We only have it once, another one is a  TransactionType which is more generic, a different type

DmitriyNoa replied:
Done

DmitriyNoa commented on packages/web/app/modules/tx/transactions/investment/sagas.ts:
```diff
@@ -160,6 +162,7 @@ function* investmentFlowGenerator({ logger }: TGlobalDependencies): Generator<an
         equityTokenName: eto.equityTokenName,
       },
     },
+    investmentType,
```
We only have it once, another one is a TransactionType which is more generic, a different type

Moejoe90 commented on packages/web/app/modules/tx/utils.ts:
```diff
@@ -31,3 +37,99 @@ export const encodeTransaction = (txData: Partial<TxData>) => ({
+  if (!decimals) {
+    switch (transaction.transactionType) {
+      case ETxSenderType.NEUR_REDEEM:
+      case ETxSenderType.USER_CLAIM:
```
can you only claim euros? 
I think you can also claim `Ether` thus this is not correct that the decimals for this transaction is always `0` 

Also I would add this logic here `startClaimGenerator` as an additional value have a look at how `tokenTransferFlowGenerator` does. 

This also includes `NEUR_REDEEM` then you don't need `getPendingTransactionDecimals`

DmitriyNoa replied:
I added decimals to all transactions type state. This function is removed

Moejoe90 commented:
This is the referenced ticket 
https://github.com/Neufund/platform-frontend/issues/3352

DmitriyNoa commented on packages/web/app/components/shared/utils.ts:
```diff
@@ -45,6 +47,27 @@ const formatEuroValueToString = (value: string) =>
+const getBlockHeight = (): number =>
+  process.env.NO_BLOCK_HIGHT ? parseInt(process.env.NO_BLOCK_HIGHT, 10) : noBlockHeightDefault;
+
+const isTransactionMined = (
```
Moved

Moejoe90 commented on packages/web/app/modules/tx/transactions/redeem/sagas.ts:
```diff
@@ -71,10 +71,12 @@ function* startNEuroRedeemGenerator(_: TGlobalDependencies): any {
   }
 
   const bankFee: string = yield select(selectBankFeeUlps);
+  const tokenDecimals = 0;
```
can you make this a constant for example `EURO_DECIMALS`

DmitriyNoa replied:
Done

Moejoe90 commented on packages/web/app/modules/tx/transactions/refund/sagas.ts:
```diff
@@ -60,12 +60,14 @@ function* startRefundGenerator(_: TGlobalDependencies, etoId: string): Generator
   const ethPrice: string = yield select(selectEtherPriceEur);
   const costUlps: string = yield select(selectTxGasCostEthUlps);
   const costEurUlps = multiplyBigNumbers([ethPrice, costUlps]);
+  const tokenDecimals = 18;
```
can you make this a contant for example `ETH_DECIMALS`

DmitriyNoa replied:
Done

DmitriyNoa commented:
@Moejoe90 Schema validators for pending transactions added.

DmitriyNoa commented on packages/web/app/components/shared/formatters/utils.ts:
```diff
@@ -18,6 +18,7 @@ export enum ENumberInputFormat {
 
 export enum ECurrency {
   NEU = "neu",
+  NEUR = "eur_t",
```
Removed

DmitriyNoa commented on packages/web/app/components/shared/transaction/Transaction.stories.tsx:
```diff
@@ -0,0 +1,82 @@
+  .add("ETO investment euro token ICBM, with icon", () => {
+    const transaction: IPendingTransaction = {
+      subType: ETransactionStatus.PENDING,
+      transactionDirection: ETransactionDirection.IN,
```
Added

Moejoe90 commented on packages/web/app/components/modals/tx-sender/investment-flow/Summary.stories.tsx:
```diff
@@ -20,13 +21,15 @@ const data = {
     estimatedReward: "40000000000000000000",
     etherPriceEur: "200",
     isIcbm: false,
+    tokenDecimals: 18,
```
please use `ETH_DECIMALS` here

Moejoe90 commented on packages/web/app/modules/tx/transactions/payout/types.ts:
```diff
@@ -0,0 +1,12 @@
+import * as YupTS from "../../../../lib/yup-ts.unsafe";
+
+export const ITokenDisbursalSchema = YupTS.object({
```
so you don't need decimals information for payout?

Moejoe90 commented on packages/web/app/modules/tx/monitor/sagas.ts:
```diff
@@ -101,37 +108,84 @@ export function* markTransactionAsPending(
-    case ETxSenderType.INVEST: {
-      const isValid = yield InvestmentAdditionalDataSchema.toYup().isValid(
-        pendingTransaction.transactionAdditionalData,
+    case ETxSenderType.USER_CLAIM: {
```
:+1:

Moejoe90 commented on packages/web/app/modules/tx/monitor/sagas.ts:
```diff
@@ -101,37 +108,84 @@ export function* markTransactionAsPending(
+  transaction: TxPendingWithMetadata,
+  senderType: ETxSenderType,
+  transactionSchema: YupSchema<any>,
+): Generator<any, any, any> {
```
why not type this `generator` return type `void`

Moejoe90 commented on packages/web/app/modules/tx/monitor/sagas.ts:
```diff
@@ -101,37 +108,84 @@ export function* markTransactionAsPending(
+export function* validateTransactionSchema(
+  transaction: TxPendingWithMetadata,
+  senderType: ETxSenderType,
+  transactionSchema: YupSchema<any>,
```
is this really `any` we actually know each schema that could pass so this type is 
`TInvestmentAdditionalData | ...`  please check how `TInvestmentAdditionalData` is done

Moejoe90 commented on packages/web/app/modules/tx/transactions/claim/types.ts:
```diff
@@ -1,7 +1,21 @@
+  costUlps: YupTS.string(),
+  tokenDecimals: YupTS.number(),
+  tokenSymbol: YupTS.string(),
+});
```
why we don't have type information for this `schema` if you check `TInvestmentAdditionalData` this could be done also here

DmitriyNoa replied:
Added

Moejoe90 commented on packages/web/app/modules/tx/transactions/redeem/types.ts:
```diff
@@ -1,8 +1,23 @@
+  bankAccount: BankAccountSchema,
+  bankFee: YupTS.string(),
+  tokenDecimals: YupTS.number(),
+});
```
why we don't have type information for this schema if you check TInvestmentAdditionalData this could be done also here

DmitriyNoa replied:
Added

Moejoe90 commented on packages/web/app/modules/tx/transactions/refund/types.ts:
```diff
@@ -7,4 +9,17 @@ export type TRefundAdditionalData = {
+  costEurUlps: YupTS.string(),
+  companyName: YupTS.string(),
+  tokenDecimals: YupTS.number(),
+});
```
why we don't have type information for this schema if you check TInvestmentAdditionalData this could be done also here

Moejoe90 commented on packages/web/app/modules/tx/transactions/token-transfer/types.ts:
```diff
@@ -10,3 +12,12 @@ export type TTokenTransferAdditionalData = {
+  tokenImage: YupTS.string(),
+  tokenSymbol: YupTS.string(),
+  tokenDecimals: YupTS.number(),
+});
```
why we don't have type information for this schema if you check TInvestmentAdditionalData this could be done also here

btw TokenTransfer and `Withdraw` share the same interface I know that as a fact

DmitriyNoa replied:
Added

Moejoe90 commented on packages/web/app/modules/tx/transactions/withdraw/types.ts:
```diff
@@ -10,3 +12,15 @@ export type TWithdrawAdditionalData = {
+  tokenImage: YupTS.string(),
+  tokenSymbol: YupTS.string(),
+  tokenDecimals: YupTS.number(),
+  investmentType: YupTS.string(),
```
why do you need `investmentType` here this is withdraw?

Withdraw and `tokenTransfer` share the same exact interface they should be one

Moejoe90 commented on packages/web/app/modules/tx/event-channel/sagas.ts:
```diff
@@ -38,8 +39,9 @@ export function* getTransactionOrThrow(
 
-  // Both requests `getTx` and `getTransactionReceipt` can end up in two separate nodes
-  const isMined = tx && tx.blockNumber && txReceipt && txReceipt.blockNumber;
+  const currentBlockNumber = yield web3Manager.getBlockNumber();
```
as we discussed this before please don't forget to add a manual delay here if `e2e` tests are running this is only the way for for us to test add an e2e test for this feature. 

please search for `NF_CYPRESS_RUN` and you will find examples

DmitriyNoa replied:
Looks like a delay is not needed. E2e can be implemented without it.

Moejoe90 replied:
rememeber we agreed to add a delay but I must have missed it in the code.
sounds good then

desfero commented on packages/web/app/components/shared/transaction/Transaction.tsx:
```diff
@@ -0,0 +1,48 @@
+        }
+      />
+    </div>
+    <Money
```
Do we need to show minus sign in case it's an outgoing transaction? It would be good to maybe add it the array of transaction history as there is no need to keep to separate `ui` for pending and history.

DmitriyNoa replied:
I think we should not add it to the transaction history array, as it's a completely different entity, and has separate design criteria.

DmitriyNoa replied:
The minus symbol is not a part of new designs https://app.zeplin.io/project/5cd04b4bbe9abf68390b8e59/screen/5d7a0d9630f33445ff019697

desfero commented on packages/web/app/modules/tx/utils.spec.ts:
```diff
@@ -0,0 +1,81 @@
+
+import { ECurrency } from "../../../../shared/src/utils/utils";
+import { ETransactionType } from "../../lib/api/analytics-api/interfaces";
+import { generalPendingTxFixture } from "../../test-e2e/pending-transactions/generalPendingTxFixture";
```
I would prefer to not import from `test-e2e` into the app as it should never use code from our e2e tests. Import from app to `test-e2e` is allowed but should be avoided and maybe in future, we will refactor the app to introduce a shared folder between app and e2e (for e.g. to keep `tid` as consts)

DmitriyNoa replied:
Fixed

desfero commented on packages/web/app/modules/tx/utils.spec.ts:
```diff
@@ -0,0 +1,81 @@
+  });
+
+  it("returns token symbol when it is set", () => {
+    delete pendingTransaction.transactionType;
```
do we need to delete if then on the next line we just assign another value?

DmitriyNoa replied:
Fixed

desfero commented on packages/web/app/components/shared/transaction/Transaction.stories.tsx:
```diff
@@ -0,0 +1,145 @@
+import { ECurrency, ETH_DECIMALS } from "../formatters/utils";
+import { Transaction } from "./Transaction";
+
+const generalPendingTxFixture = (
```
is it possible to reuse `generalPendingTxFixture` from `test-e2e`? (also in case we can use we should be `generalPendingTxFixture` outside of `test-e2e` as we shouldn't import in the app from e2e tests

DmitriyNoa replied:
Fixed

desfero commented on packages/web/app/components/shared/transaction/Transaction.stories.tsx:
```diff
@@ -0,0 +1,145 @@
+
+const generalPendingTxFixture = (
+  transactionType: ETransactionType,
+  transactionAdditionalData: Partial<{}>,
```
we it's a Partial of empty object? In general, we can generate a proper type base on `transactionType` but for story we can type it as `any` or `uknown`

desfero commented on packages/web/app/modules/tx/utils.ts:
```diff
@@ -31,3 +37,101 @@ export const encodeTransaction = (txData: Partial<TxData>) => ({
+  switch (transaction.transactionType) {
+    case ETxSenderType.INVEST:
+      return ETransactionType.ETO_INVESTMENT;
+      break;
```
do we need to use `break` here if we just `return` always?

DmitriyNoa replied:
Fixed

desfero commented on packages/web/app/modules/tx/utils.ts:
```diff
@@ -31,3 +37,101 @@ export const encodeTransaction = (txData: Partial<TxData>) => ({
+  return amount;
+};
+
+export const getPendingTransactionCurrency = (transaction: TxPendingWithMetadata): ECurrency => {
```
can we merge `getPendingTransactionAmount` and `getPendingTransactionCurrency` as these two are coupled together (amount and currency) and should be implemented together as logic often is the same IMO

DmitriyNoa replied:
It might be difficult to merge this 2. It will be better to unify the data model of the pending transaction as described [here](https://github.com/Neufund/platform-frontend/issues/3967) and get rid of these 2 utility functions.

DmitriyNoa commented on packages/web/app/components/shared/transaction/Transaction.module.scss:
```diff
@@ -0,0 +1,26 @@
+}
+
+.icon {
+  padding: .5rem 1rem .5rem 1.5rem;
```
It is not a part of the transaction history table - but a self-standing component, it just happened to be used on top of the history, but potentially could be used it other places. If I remove the padding, in this case, it wouldn't match the design specs.

DmitriyNoa commented on packages/web/app/components/wallet/transactions-history/TransactionsHistory.tsx:
```diff
@@ -35,10 +39,16 @@ const TransactionListLayout: React.FunctionComponent<TStateProps & TDispatchProp
-    {transactionsHistoryPaginated.transactions && (
-      <div className={styles.wrapper}>
+    {pendingTransaction && (
+      <div className={styles.pendingTransactionWrapper}>
```
I think we should not treat pending transactions as a part of the transaction history list and component, as it is a separate entity.

DmitriyNoa commented on packages/web/app/modules/tx/event-channel/utils.ts:
```diff
@@ -0,0 +1,25 @@
+
+const noBlockHeightDefault = 0;
+const getBlockHeight = (): number =>
+  process.env.NO_BLOCK_HIGHT ? parseInt(process.env.NO_BLOCK_HIGHT, 10) : noBlockHeightDefault;
```
NO_BLOCK_HEIGHT - was suggested by @Moejoe90 . What do you think?

Moejoe90 replied:
We can change it to how it makes more sense I have no strong feelings. 
`NO` was coming from the idea of number `BLOCK_HIGHT_NO`  for example is the blok hight number.

DmitriyNoa replied:
No need for delay. It's possible to implement e2e tests without it

DmitriyNoa replied:
I renamed it as discussed.

DmitriyNoa commented on packages/web/app/components/wallet/transactions-history/TransactionsHistory.tsx:
```diff
@@ -35,10 +39,16 @@ const TransactionListLayout: React.FunctionComponent<TStateProps & TDispatchProp
+        <Transaction icon={<PendingTransactionImage />} transaction={pendingTransaction} />
+      </div>
+    )}
+    <div className={styles.wrapper}>
```
I added the Transaction component here as a part of the same div, previously it was used only for Transaction history. Maybe we can add a condition to show the div only if pending or transaction history exists.

DmitriyNoa commented on packages/web/app/modules/tx/event-channel/utils.ts:
```diff
@@ -0,0 +1,25 @@
+): boolean => {
+  const blockHeight = getBlockHeight();
+
+  const isMined =
```
This will not work as tx and txReciept could be null and TS will complain about it, so we have to check if tx and txReciept exist first.

DmitriyNoa commented on packages/web/app/test-e2e/pending-transactions/SupportForPendingTransactionsInHeader.spec.e2e.io.ts:
```diff
@@ -136,6 +136,8 @@ describe("Pending Transactions In Header", () => {
         closeWhen: "success",
       });
 
+      cy.wait(3000);
```
This is a leftover from local testing. Removed

DmitriyNoa commented on packages/web/app/modules/tx/monitor/sagas.ts:
```diff
@@ -101,37 +109,86 @@ export function* markTransactionAsPending(
   return transactionTimestamp;
 }
 
+export function* validateTransactionSchema(
```
Good idea will check it

Moejoe90 commented on packages/web/app/modules/tx/utils.ts:
```diff
@@ -31,3 +37,101 @@ export const encodeTransaction = (txData: Partial<TxData>) => ({
   value: addHexPrefix(new BigNumber((txData.value && txData.value.toString()) || "0").toString(16)),
   data: addHexPrefix(txData.data || EMPTY_DATA),
 });
+
```
also Please don't forget to add a `TODO` comment that links to the combining interface PR. 
As we need to get rid of these utils as soon as possible

DmitriyNoa replied:
it's already there, in the selector

DmitriyNoa commented on packages/web/app/lib/yup-ts.unsafe.tsx:
```diff
@@ -56,7 +56,7 @@ class YTS<T> {
 }
 
-class ObjectYTS<T> extends YTS<TypeOfProps<T>> {
+export class ObjectYTS<T> extends YTS<TypeOfProps<T>> {
```
Removed

DmitriyNoa commented on packages/web/app/components/wallet/transactions-history/TransactionsHistory.tsx:
```diff
@@ -7,23 +7,27 @@ import { branch, compose, renderComponent } from "recompose";
 import { Panel } from "../../shared/Panel";
 import { ENewTableTheme, NewTableRow, Table } from "../../shared/table";
+import { Transaction } from "../../shared/transaction/Transaction";
 import { TransactionData } from "../../shared/TransactionData";
```
Moved

DmitriyNoa commented on packages/web/app/modules/tx/monitor/selectors.ts:
```diff
@@ -14,6 +23,42 @@ export const selectPlatformPendingTransaction = (
+  let transaction = null;
+
+  if (pending && pending.transactionStatus === ETxSenderState.MINING) {
+    transaction = {
```
Changed

DmitriyNoa commented on packages/web/app/components/shared/formatters/utils.ts:
```diff
@@ -303,6 +303,9 @@ export const stripNumberFormatting = (value: string) => {
 };
 
+export const EURO_DECIMALS = 0;
+export const ETH_DECIMALS = 18;
```
Fixed

DmitriyNoa commented:
@desfero I added missing e2e tests. Your question regarding delay - as for now, as soon as the transaction is mined and removed from pending we are making a request for new transaction history. So the delay is difficult to decrease.

What we can do maybe is prepend a new transaction to the history array, so it will be displayed there until the history will be refreshed. The tricky part there is totally different data model and it would be better to do in a different ticket. Please tell me what do you think?

Moejoe90 commented on packages/web/app/modules/tx/monitor/sagas.ts:
```diff
@@ -195,7 +168,11 @@ export function* updatePendingTxs({
   try {
     // If there is a pending transaction, check if schema is valid
-    if (apiPendingTx.pendingTransaction) {
+    // SIGN_INVESTMENT_AGREEMENT send an empty data, so no validation is required
```
why then you dont't just specify an empty schema for `SIGN_INVESTMENT_AGREEMENT` maybe in the future we will add some kind of data

DmitriyNoa replied:
Created an empty schema for this, the condition is removed

Moejoe90 replied:
can you remove in this case the comment?

Moejoe90 commented on packages/web/app/modules/tx/transactions/types.ts:
```diff
@@ -0,0 +1,47 @@
+  | TInvestorAcceptPayouAdditionalData
+  | TNeurRedeemAdditionalData
+  | TokenTransferAdditionalData
+  | InvestorRefundAdditionalData;
```
you are missing some transactions here please check them this includes `unlock` for example please add them all

DmitriyNoa replied:
added all types

Moejoe90 commented on packages/web/app/modules/tx/transactions/nominee/sign-agreement/types.ts:
```diff
@@ -11,3 +12,8 @@ export interface IAgreementContractAndHash {
+export const TokenAgreementContractSchema = YupTS.object({
+  contract: YupTS.string(),
+  currentAgreementHash: YupTS.string(),
+});
```
why is there no type informstion

DmitriyNoa replied:
added

Moejoe90 commented on packages/web/app/modules/tx/transactions/unlock/types.ts:
```diff
@@ -1,5 +1,13 @@
+  etherNeumarksDue: YupTS.string(),
+  lockedEtherUnlockDate: YupTS.string(),
+  lockedEtherBalance: YupTS.string(),
+});
```
why is there no type information

DmitriyNoa replied:
added

Moejoe90 commented on packages/web/app/modules/tx/transactions/upgrade/types.ts:
```diff
@@ -1,5 +1,9 @@
 };
+export const UpgradeAdditionalDataSchema = YupTS.object({
+  tokenType: YupTS.string(),
+});
```
why is there no type information

DmitriyNoa replied:
added

DmitriyNoa commented:
@desfero Actually I was able to find how to reload transactions history, so it is refreshed immediately

desfero commented on packages/web/app/modules/tx/monitor/sagas.ts:
```diff
@@ -254,6 +173,7 @@ export function* updatePendingTxs({
     }
 
     yield put(actions.txMonitor.setPendingTxs(apiPendingTx));
+    yield actions.txHistory.loadTransactions();
```
this one won't work as you need to `yield put(actions.txHistory.loadTransactions());`

desfero commented on packages/web/app/test-e2e/pending-transactions/PendingTransactionsComponent.spec.e2e.io.ts:
```diff
@@ -0,0 +1,38 @@
+import { Q18 } from "@neufund/shared";
```
Maybe we can rename file to `PendingTransactoinInHistory` to clearly state that it's a pending transaction in tx history?

desfero commented on packages/web/app/test-e2e/pending-transactions/PendingTransactionsComponent.spec.e2e.io.ts:
```diff
@@ -0,0 +1,38 @@
+    }).then(({ address }) => {
+      cy.saveLocalStorage();
+      userAddress = address;
+      sendEth("DEPLOYER", address, Q18.mul(amount));
```
Is there a reason to feed the account with ether?

desfero commented on packages/web/app/modules/tx/transactions/types.ts:
```diff
@@ -1,8 +1,55 @@
 
-export type TransactionAdditionalData =
-  | TInvestmentAdditionalData
+export const typeToSchema: Record<ETxSenderType, Schema<unknown> | null> = {
```
no need for `null` as a value IMO, the schema is always provided, maybe it's a leftover from previous refactoring.

pavelneu commented on packages/web/app/components/modals/tx-sender/investment-flow/InvestmentTransactionDetails.spec.tsx:
```diff
@@ -24,10 +25,12 @@ describe("InvestmentTransactionDetails", () => {
     investmentEur: "44000000000000000000",
     gasCostEth: "1200000000000000",
     equityTokens: "73",
+    tokenDecimals: 18,
```
missed variable definition
should it be ETH_DECIMALS ?

pavelneu commented on packages/web/app/components/shared/transaction/Transaction.stories.tsx:
```diff
@@ -0,0 +1,58 @@
+
+    return <Transaction transaction={transaction} />;
+  })
+  .add("Invest EHT", () => {
```
```suggestion
  .add("Invest ETH", () => {
```

DmitriyNoa commented:
@pavelneu @glupiawymowka I fixed the Nominee Sign THA. Please check

DmitriyNoa commented:
@pavelneu "When Investor clicks on Redistribute Payout, the name of transaction is not correct: Received instead of Redistribute Payout from NEU" - **is fixed**, please check.
